packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -295,7 +295,20 @@ function commitHookEffectList(
       if ((effect.tag & mountTag) !== NoHookEffect) {
         // Mount
         const create = effect.create;
-        const destroy = create();
+        let destroy = create();
+        if (typeof destroy !== 'function') {
+          if (__DEV__ && destroy !== null && destroy !== undefined) {
+            warningWithoutStack(
+              false,
+              'useEffect function must return a cleanup function or nothing.%s%s',
+              typeof destroy.then === 'function'
+                ? ' Promises and async functions are not supported.'
+                : '',
+              getStackByFiberInDevAndProd(finishedWork),
+            );
+          }
+          destroy = null;
```

Comment:
In production the branch `if (typeof destroy !== 'function') {` will still be present with `destroy = null`. Is it worth keeping it since the next line checks the type again?

Reply:
Thanks for the catch, I intended to remove the conditional from the next line.